### PR TITLE
Transfer between grids functionality for subgraph parallelism

### PR DESCRIPTION
**Pull request by aj-prime**
PR introduces transfer between grids functionality mainly for `STAR,VC` distributions. These functions are designed for subgraph parallelism support in LBANN. Mainly there are four types of functions 

1. Allreduce to sum the output of subgrids to parent gird (Sum layer)
2. Broadcast to copy data from parent to subgrids (Split layer)
3. Scatter to scatter data from parent to subgrids (Slice layer)
4. Gather to concatenate data of subgrids to parent grid (Concatenate layer)

Basic TranslateBetweenGrids function for `STAR,VC` distribution 

Four tests are also added for the performance evaluation

### CODE REVIEW

**Comment by benson31** [comment]
I haven't finished a full review but two preliminary comments:

1. There are several compiler warnings about unused variables. Please resolve by using these variables, commenting them out, or deleting them.
2. There are undefined references for FP16-enabled builds. Please add ETI for the `{cpu,gpu}_half_type` or, if these types are not meant to be supported, add overloads for them that just throw exceptions.

File: include/El/blas_like/level1/Copy/TranslateBetweenGrids.hpp [hunk 1]
```diff
@@ -21,8 +21,8 @@ void TranslateBetweenGrids
 {
     EL_DEBUG_CSE
 
-    if (D1 != Device::CPU)
-        LogicError("TranslateBetweenGrids: Device not implemented.");
+    //if (D1 != Device::CPU)
+        //LogicError("TranslateBetweenGrids: Device not implemented.");
```
Comment (timmoon10) [inline include/El/blas_like/level1/Copy/TranslateBetweenGrids.hpp]:
```suggestion
```

File: include/El/blas_like/level1/Copy/TranslateBetweenGrids.hpp [hunk 2]
```diff
@@ -40,9 +40,9 @@ void TranslateBetweenGrids
 {
     EL_DEBUG_CSE
 
-    if (D1 != Device::CPU)
-        LogicError("TranslateBetweenGrids<MC,MR,ELEMENT>: "
-                   "Device not implemented.");
+    //if (D1 != Device::CPU)
+        //LogicError("TranslateBetweenGrids<MC,MR,ELEMENT>: "
+                   //"Device not implemented.");
```
Comment (timmoon10) [inline include/El/blas_like/level1/Copy/TranslateBetweenGrids.hpp]:
```suggestion
```

File: include/El/blas_like/level1/Copy/TranslateBetweenGrids.hpp [hunk 3]
```diff
@@ -81,12 +81,15 @@ void TranslateBetweenGrids
...
     const bool inAGrid = A.Participating();
+
     if(!inBGrid && !inAGrid)
         return;
 
     const Int maxSendSize =
       (m/(colStrideA*numColSends)+1) * (n/(rowStrideA*numRowSends)+1);
 
+    //std::printf("colStrideA %d rowStrideA %d\n",colStrideA,rowStrideA);
+
```
Comment (timmoon10) [inline include/El/blas_like/level1/Copy/TranslateBetweenGrids.hpp]:
```suggestion
```

File: include/El/blas_like/level1/Copy/TranslateBetweenGrids.hpp (hunk 1 above)
Comment (benson31) [inline include/El/blas_like/level1/Copy/TranslateBetweenGrids.hpp]:
```suggestion
```

File: include/El/blas_like/level1/Copy/TranslateBetweenGrids.hpp (hunk 2 above)
Comment (benson31) [inline include/El/blas_like/level1/Copy/TranslateBetweenGrids.hpp]:
```suggestion
```

File: include/El/blas_like/level1/Copy/TranslateBetweenGrids.hpp (hunk 3 above)
Comment (benson31) [inline include/El/blas_like/level1/Copy/TranslateBetweenGrids.hpp]:
```suggestion
```

File: include/El/blas_like/level1/Copy/TranslateBetweenGrids.hpp [hunk 4]
```diff
@@ -136,6 +139,8 @@ void TranslateBetweenGrids
     Int recvRow = 0; // avoid compiler warnings...
     if(inAGrid)
         recvRow = Mod(Mod(colRankA-colAlignA,colStrideA)+colAlignB,colStride);
+
+
```
Comment (benson31) [inline include/El/blas_like/level1/Copy/TranslateBetweenGrids.hpp]:
```suggestion
```

File: include/El/blas_like/level1/Copy/TranslateBetweenGrids.hpp [hunk 5]
```diff
@@ -241,129 +247,4016 @@ void TranslateBetweenGrids
 }
 
 
+
+
```
Comment (benson31) [inline include/El/blas_like/level1/Copy/TranslateBetweenGrids.hpp]:
```suggestion
```

File: include/El/blas_like/level1/Copy/internal_decl.hpp [hunk 6]
```diff
@@ -38,10 +38,178 @@ void Translate
...
+  int splitDim,  
+  mpi::Comm const& gatherComm, 
+  SyncInfo<D1> & syncGeneral);
+
+
+
+// void TranslateBetweenGridsBroadcast
+// (DistMatrix<double,STAR,VC,ELEMENT,Device::CPU> const& A,
+//   std::vector<DistMatrix<double,STAR,VC,ELEMENT,Device::CPU>>& B_Vector);
+
```
Comment (benson31) [inline include/El/blas_like/level1/Copy/internal_decl.hpp]:
```suggestion
```

File: include/El/blas_like/level1/Copy/TranslateBetweenGrids.hpp [hunk 7]
```diff
@@ -241,129 +247,4016 @@ void TranslateBetweenGrids
...
+    if(rankA == 0)
+        mpi::Wait(sendRequest);
+//#endif // EL_TRANSLATE_BETWEEN_GRIDS_REENABLE__
+}
+
+} // namespace copy
+} // namespace El
+
+
+//template void TranslateBetweenGridsBroadcast <double, Device::GPU,Device::GPU> (DistMatrix<double,STAR,VC,ELEMENT,Device::GPU> const& , std::vector<DistMatrix<double,STAR,VC,ELEMENT,Device::GPU>>& );
```
Comment (benson31) [inline include/El/blas_like/level1/Copy/TranslateBetweenGrids.hpp]:
```suggestion
```

File: include/El/blas_like/level1/Copy/TranslateBetweenGrids.hpp [hunk 8]
```diff
@@ -241,129 +247,4016 @@ void TranslateBetweenGrids
...
+template<typename T, Device D1, Device D2>
+void TranslateBetweenGrids
+(const DistMatrix<T,STAR,STAR,ELEMENT,D1>& A,
+  DistMatrix<T,STAR,STAR,ELEMENT,D2>& B)
+{
+    EL_DEBUG_CSE;
+    //LogicError("TranslateBetweenGrids is no longer supported. "
+               //"If you have reached this message, please open "
+               //"an issue at https://github.com/llnl/elemental.");
+//#ifdef EL_TRANSLATE_BETWEEN_GRIDS_REENABLE__
```
Comment (benson31) [inline include/El/blas_like/level1/Copy/TranslateBetweenGrids.hpp]:
```suggestion
```

File: include/El/blas_like/level1/Copy/internal_decl.hpp (hunk 6 above)
Comment (timmoon10) [inline include/El/blas_like/level1/Copy/internal_decl.hpp]:
```suggestion
```

File: include/El/blas_like/level1/Copy/TranslateBetweenGrids.hpp [hunk 9]
```diff
@@ -241,129 +247,4016 @@ void TranslateBetweenGrids
...
+
+        Synchronize(syncGeneral);
+        mpi::AllReduce( sendBuf, recvBuf, maxSendSize, mpi::SUM, viewingCommA,syncInfoB);
+
+
+        if(inAGrid)
+        {
+            if(require_data[localRankB])
+            {
+                int sendWidth = int(n / rowLCM);
```
Comment (timmoon10) [inline include/El/blas_like/level1/Copy/TranslateBetweenGrids.hpp:363]: I don't think we're properly handling the case where `n` doesn't evenly divide `rowLCM`:
```suggestion
                int sendWidth = (n + rowLCM - 1) / rowLCM;
                if (localRankB > n % rowLCM)
                {
                    --sendWidth;
                }
```

**Review by benson31 — COMMENTED** [review]
Just some dead code/whitespace cleanup at this point. I want to split the 4000 line file. Let's talk more at the meeting this afternoon.

File: include/El/blas_like/level1/Copy/TranslateBetweenGrids.hpp [hunk 10]
```diff
@@ -241,129 +247,4016 @@ void TranslateBetweenGrids
...
-#ifdef EL_TRANSLATE_BETWEEN_GRIDS_REENABLE__
-    const Int height = A.Height();
-    const Int width = A.Width();
-    B.Resize(height, width);
+    //<T,STAR,VC,ELEMENT,D2>
+    /*
+    This function is specific to the LBANN with implementation for specific cases
+    Subgrids in B_vector are assumed to be subset of resources in A grid 
+    Sum the output from different subgrids to parent grid
+    Basic allreduce implementation without overlapping communication
```
Comment (timmoon10) [inline include/El/blas_like/level1/Copy/TranslateBetweenGrids.hpp]:
```suggestion
    Logically, the values in B_vector are summed together and copied to A. 
    This function is specific to the LBANN with implementation for specific cases.
    
    Subgrids in B_vector are assumed to evenly divide the grid in A.
    This is a basic allreduce implementation with no overlapped communication.
```

File: include/El/blas_like/level1/Copy/TranslateBetweenGrids.hpp [hunk 11]
```diff
@@ -241,129 +247,4016 @@ void TranslateBetweenGrids
...
+{
+    //<T,STAR,VC,ELEMENT,D2>
+    /*
+    This function is specific to the LBANN with implementation for specific cases
+    Subgrids in B_vector are assumed to be subset of resources in A grid 
+    Sum the output from different subgrids to parent grid
+    This function requires comm object 
+    MPI_COMM_SPLIT is an expensive operation on GPU
+    LBANN uses this function in subgraph parallelism
+
```
Comment (timmoon10) [inline include/El/blas_like/level1/Copy/TranslateBetweenGrids.hpp]:
```suggestion
    Logically, the values in B_vector are summed together and copied to A.
    This function is specific to LBANN sub-graph parallelism with implementations for
    specific cases.
    
    Subgrids in B_vector are assumed to evenly divide the grid in A.
    This is a segmented allreduce implementation that requires a
    a communicator over corresponding processes in the subgrids.
```

File: include/El/blas_like/level1/Copy/TranslateBetweenGrids.hpp [hunk 12]
```diff
@@ -241,129 +247,4016 @@ void TranslateBetweenGrids
...
 
-    // Receive on the root of B's matrix's grid and then broadcast
-    // over the owning communicator
-    if(B.Participating())
+    if(inAGrid)
     {
-        if (bcastBuffer.size() != size_t(height*width))
-            RuntimeError("TranslateBetweenGrids: Bad bcastBuffer size!");
-        if(rankB == 0)
+        int sendWidth = int(n / rowLCM);
```
Comment (timmoon10) [inline include/El/blas_like/level1/Copy/TranslateBetweenGrids.hpp:480]: Need to handle case where `n` doesn't cleanly divide `rowLCM`.

File: include/El/blas_like/level1/Copy/TranslateBetweenGrids.hpp [hunk 13]
```diff
@@ -241,129 +247,4016 @@ void TranslateBetweenGrids
...
+  std::vector<std::unique_ptr<AbstractDistMatrix<T>>>& B_Vector)
+{
+    //<T,STAR,VC,ELEMENT,D2>
+    /*
+    This function is specific to the LBANN with implementation for specific cases
+    Subgrids in B_vector are assumed to be subset of resources in A grid 
+    Sum the output from different subgrids to parent grid
+    Basic allreduce implementation without overlapping communication
+    This function does not require comm object 
+    MPI_COMM_SPLIT is an expensive operation on GPU
```
Comment (timmoon10) [inline include/El/blas_like/level1/Copy/TranslateBetweenGrids.hpp]:
```suggestion
    Logically, the values in B_vector are summed together and copied to A.
    This function is specific to LBANN sub-graph parallelism with implementations for
    specific cases.
    
    Subgrids in B_vector are assumed to evenly divide the grid in A.
    This is a segmented allreduce implementation that internally splits a communicator.
    Note that communicator splitting is expensive on GPU.
```

File: include/El/blas_like/level1/Copy/TranslateBetweenGrids.hpp [hunk 14]
```diff
@@ -241,129 +247,4016 @@ void TranslateBetweenGrids
...
+                1, B->LDim(),
+                sendBuf, 1, mLocB, syncGeneral);
+
+    }
+    //Synchronize(syncGeneral);
+    mpi::AllReduce( sendBuf, recvBuf, maxSendSize, mpi::SUM, allreduceComm,syncGeneral);
+
+    if(inAGrid)
+    {
+        int sendWidth = int(n / rowLCM);
```
Comment (timmoon10) [inline include/El/blas_like/level1/Copy/TranslateBetweenGrids.hpp:599]: Need to handle case where `n` doesn't cleanly divide `rowLCM`.

**Review by timmoon10 — COMMENTED** [review]
I've finished working through the `TranslateBetweenGridsAllreduceBasic` implementations. My main suggestions here are to make our assumptions explicit and to handle correctness issues when the matrix width doesn't cleanly divide the grid size.

File: include/El/blas_like/level1/Copy/TranslateBetweenGrids.hpp [hunk 15]
```diff
@@ -241,129 +247,4016 @@ void TranslateBetweenGrids
...
+        //Aluminum infer comm from sync object 
+        
+
+       
+        mpi::Scatter((T *)transposedMatrix.Buffer(), sendCounts, (T *)recvTransposedMatrix.Buffer(), sendCounts, localDataRankA, ScatterComm,
+               syncGeneral);
+
+        Synchronize(syncGeneral);
+
+        int sendWidth = int(n / rowLCM);
```
Comment (timmoon10) [inline include/El/blas_like/level1/Copy/TranslateBetweenGrids.hpp:789]: Doesn't handle case where `n` isn't cleanly divisible.

**Review by timmoon10 — APPROVED** [review]
This PR seems reasonable for nicely grid sizes and matrix dimensions, although I have to admit I haven't been able to grok a lot of the index management. We will run into trouble if the sub-grids don't evenly partition the parent grid or if the matrix width is not nice (e.g. the last mini-batch in an epoch). That said, this should only affect the sub-grid parallelism work. We should merge this in so we don't get stuck.

**Review by benson31 — APPROVED** [review]
A few issues.

1. The tests are not listed in the `tests/core/CMakeLists.txt`.
2. The tests use `cudaDeviceSynchronize()`. It should use `El::gpu::SynchronizeDevice()`.

Once these are addressed, this is ok. I share Tim's concern about the generality with respect to grid sizes.

File: include/El/blas_like/level1/Copy/TranslateBetweenGrids.hpp (hunk 12 above)
Comment (aj-prime) [reply include/El/blas_like/level1/Copy/TranslateBetweenGrids.hpp:480]: Shall I update all the functions to handle this case or insert a check?

Comment (aj-prime) [reply include/El/blas_like/level1/Copy/TranslateBetweenGrids.hpp:480]: I need to go over all the functions and understand their implementation if we want to support this case.

**Comment by aj-prime** [comment]
Hi @benson31 and @timmoon10,

I have taken in all the comments and resolved the issues. Sanity tested the tests on a single node.

**Review by timmoon10 — APPROVED** [review]

**Pull request merged**